File: Groak_iOS/Groak/Groak/RestaurantObjects/Restaurant.swift
//
//  Restaurant.swift
//  Groak
//
//  Created by Vishisht Tiwari on 11/6/19.
//  Copyright © 2019 Groak. All rights reserved.
//
//  Restaurant class used to refer to a restaurant

import UIKit
import Firebase

internal class Restaurant {
    var reference: DocumentReference?
    var name: String
    var type: [String]
    var logo: String
    var salesTax: Double
    var address: [String: Any]
    var latitude: Double
    var longitude: Double
    var maximumOccupancy: Int
    var currentOccupancy: Int
    var occupancy: [String: Int]
    var covidGuidelines: String
    var covidMessage: String
    
    init() {
        reference = nil
        name = ""
        type = []
        logo = ""
        salesTax = -1
        address = [:]
        latitude = -1000
        longitude = -1000
        maximumOccupancy = 10
        currentOccupancy = 0
        occupancy = [:]
        covidGuidelines = ""
        covidMessage = ""
    }
    
    init(_ temp: String) {
        let db = Firebase.db;
        
        reference = db.collection("/restaurants").document("361mKaKkOCW73UYueUs6eSHhOC92")
        name = "The Yellow Chilli"
        type = []
        logo = ""
        salesTax = -1
        address = [:]
        latitude = -1000
        longitude = -1000
        maximumOccupancy = 10
        currentOccupancy = 0
        occupancy = [:]
        covidGuidelines = ""
        covidMessage = ""
    }
    
    init(restaurant: [String: Any]) {
        reference = restaurant["reference"] as? DocumentReference
        name = restaurant["name"] as? String ?? ""
        type = restaurant["type"] as? [String] ?? []
        logo = restaurant["logo"] as? String ?? ""
        salesTax = restaurant["salesTax"] as? Double ?? -1
        address = restaurant["address"] as? [String: Any] ?? [:]
        latitude = (restaurant["address"] as? [String: Any] ?? [:])["latitude"] as? Double ?? -1000
        longitude = (restaurant["address"] as? [String: Any] ?? [:])["longitude"] as? Double ?? -1000
        maximumOccupancy = restaurant["maximumOccupancy"] as? Int ?? 10
        currentOccupancy = restaurant["currentOccupancy"] as? Int ?? 0
        occupancy = restaurant["occupancy"] as? [String: Int] ?? [:]
        covidGuidelines = restaurant["covidGuidelines"] as? String ?? ""
        covidMessage = restaurant["covidMessage"] as? String ?? ""
    }
    
    func success() -> Bool {
        if (reference == nil) {
            return false;
        } else if (name.count == 0) {
            return false;
        }  else if (salesTax < 0) {
           return false;
       } else if (latitude == -1000) {
            return false;
        } else if (longitude == -1000) {
            return false;
        }
        return true;
    }
    
    var description : String {
        var str = "Restaurant Name: \(name)\n"
        str += "Restaurant Type: \(type)\n"
        str += "Restaurant Logo: \(logo)\n"
        str += "Restaurant Latitude: \(latitude)\n"
        str += "Restaurant Longitude: \(longitude)"
        
        return str;
    }
}

File: Groak_android/app/src/main/java/com/groak/groak/restaurantobject/Restaurant.java
package com.groak.groak.restaurantobject;

import com.google.firebase.firestore.DocumentReference;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Restaurant {
    private DocumentReference reference;
    private String name;
    private ArrayList<String> type;
    private String logo;
    private long salesTax;
    private HashMap<String, Object> address;
    private double latitude;
    private double longitude;
    private long maximumOccupancy;
    private long currentOccupancy;
    private HashMap<String, Integer> occupancy;
    private String covidGuidelines;
    private String covidMessage;

    public Restaurant(Map<String, Object> map) {
        this.reference = (DocumentReference)map.get("reference");
        this.name = (String)map.get("name");
        this.type = (ArrayList<String>)map.get("type");
        this.logo = (String)map.get("logo");
        this.salesTax = (long)map.get("salesTax");
        this.address = (HashMap<String, Object>)map.get("address");
        this.latitude = (double)address.get("latitude");
        this.longitude = (double)address.get("longitude");
        this.maximumOccupancy = (long)map.get("maximumOccupancy");
        this.currentOccupancy = (long)map.get("currentOccupancy");
        this.occupancy = (HashMap<String, Integer>)map.get("occupancy");
        this.covidGuidelines = (String)map.get("covidGuidelines");
        this.covidMessage = (String)map.get("covidMessage");
    }

    public DocumentReference getReference() {
        return reference;
    }
    public String getName() {
        return name;
    }
    public ArrayList<String> getType() {
        return type;
    }
    public String getLogo() {
        return logo;
    }
    public long getSalesTax() {
        return salesTax;
    }
    public HashMap<String, Object> getAddress() {
        return address;
    }
    public double getLatitude() {
        return latitude;
    }
    public double getLongitude() {
        return longitude;
    }
    public long getMaximumOccupancy() {
        return maximumOccupancy;
    }
    public long getCurrentOccupancy() {
        return currentOccupancy;
    }
    public HashMap<String, Integer> getOccupancy() {
        return occupancy;
    }
    public String getCovidGuidelines() {
        return covidGuidelines;
    }
    public String getCovidMessage() {
        return covidMessage;
    }

    public String toString() {
        String str = "";
        str += "Reference: " + reference + "\n";
        str += "Name: " + name + "\n";
        str += "Address: " + address + "\n";
        str += "Latitude: " + latitude + "\n";
        str += "Longitude: " + longitude + "\n";
        str += "Occupancy: " + occupancy + "\n";
        str += "Sales Tax: " + salesTax + "\n";
        return str;
    }
}
